[READY] Light attenuation based system for FoV and dynamic lighting.

kevingranade opened the pull request:
This gets rid of a bunch of heuristics for determining sight range, replacing them with one system, which is that each square on the map absorbs a fraction of the light passing through it.  This applies both to vision and to light.  The sole exception is that all "outside" squares are lit by whatever amount is called for by the weather.
Fixes #3002
- [x] Fix Night vision traits.
- [x] Fix Night vision and infravision CBMs.
- [x] Adjust light level at night to allow for short-ranged sight. (leave total darkness in for underground)
- [x] Light level fixes
  - [x] Fix light level of Shocker brutes (and similar low-luminosity monsters).
  - [x] Make Sparks from electricity glow properly.
  - [x] Reduce illumination from small fires and make sure they taper off.
  - [x] Flashlights and turrets give tons of light. Single tile of raging fire is basically as bright as daylight.
  - [x] Moonlight provides tons of light to the point where by day 8 of spring it's about 80% of the strength of a flashlight.
- [x] Make sure lighting is uniform across x and y.
- [x] Make sure tiles marked as seen or unseen are treated consistently by everything.
- [x] Investigate slowdown from large fires.
- [x] Fix issue where distant illumination lights up a line between the illumination and the player.
- [x] Fix issue where walls are invisible to nightvision.
- [ ] Only illuminate walls if the light source is on the same side of the wall as the player. (possibly another PR if the fix isn't simple enough)
- [x] Fix reading being allowed at night with no illumination.
- [x] Fix illumination from fields being persistent after the field disappears.
- [x] More performance work.
- [x] Make sources of illumination visible from much further away with sm.
- [x] Fix illuminated monsters illuminating just their squares but not themselves.
- [ ] "jagged" light borders?
- [x] Reintroduce restricted vision from weather by adding global lowered transparency in various weather conditions.
- [x] Update crafting menu to reflect new brightness values.
- [x] Weather effects don't show up in nighvision-visible areas
- [x] Near areas with flickering light such as fires, nightvision constantly flickers on and off
- [x] Getting boomer bile'd doesn't prevent redrawing of tiles due to weather effects (and presumably SCT)

kevingranade commented:
Whoops, missed some warnings, cleaning up now.

Rivet-the-Zombie commented:
Very nice!

Treah commented:
Does this fix the problem with light not transmitting though open windows and doors? Or is this some other fix?

kevingranade commented:
If you're asking about what I think you are, yes.  The the code that projected daylight into buildings was broken.

Coolthulhu commented:
Is this ready for testing?

kevingranade commented:
Yes.

Coolthulhu commented:
Certainly bugs:
* Night vision traits (even debug vision) don't work
* Night vision and infravision CBMs don't work
* During the night the darkness is so complete that player can only see adjacent creatures, not even adjacent tiles.
* Shocker brutes (and similar low-luminosity monsters) don't shine at all
* Sparks from electricity don't shine and are only barely visible (1 tile) beyond the edge of light
* (Small) fire allows crafting/reading/etc. from a really good range. Then it very quickly decays to darkness, without a even a single "shady" step in between.
* Light effects differ in x and y. On y axis, some weird shadows can spawn. Those look as if something was casting shadows rather than light. They happen even when the shadowed tiles are lit up by other light sources other than the shaded one. Can be seen by setting a building on fire and standing diagonally to it.
* Sometimes tiles are badly marked as unseen - e'x'amining them shows them as seen (and describes what is on the tile), while they're drawn as gray '#'. This can happen even when the tile is producing light (for example, has a raging fire on it).
* Small fire, flashlights and turrets give tons of light. Single tile of raging fire is basically as bright as daylight.
* Moonlight provides tons of light to the point where by day 8 of spring it's about 80% of the strength of a flashlight.

Issues that might not be bugs:
* First few turns of dawn are a transition from no light at all to flashlight strength light. The transition is quite jarring at first. Change in light levels happens every 2-3 turns at first, increasing by 2-3 tiles. By the end of the dawn it smooths out nicely, though.
* Drawing large fires looks significantly slower than before
* When a distant light-producing object exists, player can see everything in a rectangle from own position to the light-producing object. Depending on angle, this rectangle can be a 1 tile line or a large area
* Weak light-producing objects (glowstick, oil lamp) on the ground are only visible from about 2x the range of light they provide.

It looks like either there is a rounding error or some of the formulas only make sense for large numbers.

kevingranade commented:
Sorry I gave the go-ahead with this many issues, I figured I'd missed some things and they'd have to get fixed later, but I didn't think it was this much.

> During the night the darkness is so complete that player can only see adjacent creatures, not even adjacent tiles.

This was accidental, but I'd actually prefer to keep it this way if it's total darkness, I'll see if I can get the lowest outside light level set to something that gives one square of vision, but leave caves and indoors at night pitch black.

> First few turns of dawn are a transition from no light at all to flashlight strength light. The transition is quite jarring at first. Change in light levels happens every 2-3 turns at first, increasing by 2-3 tiles. By the end of the dawn it smooths out nicely, though.

I'll look at it, but this has never been particularly good, so if it's too much of a mess I'm not going to bother.

> Weak light-producing objects (glowstick, oil lamp) on the ground are only visible from about 2x the range of light they provide.

This is mostly intentional, but I can see about tweaking it.

> It looks like either there is a rounding error or some of the formulas only make sense for large numbers.

It might have issues with the single-digit numbers, also the threshold for lit at all is higher now.

Coolthulhu commented:
> Sorry I gave the go-ahead with this many issues, I figured I'd missed some things and they'd have to get fixed later, but I didn't think it was this much.

No problem, I had to test all of those things (fires, creatures, items etc.) anyway. Documenting bugs takes way less time than testing.

> This was accidental, but I'd actually prefer to keep it this way if it's total darkness

The problem is, the rest of the game doesn't handle it nicely. You can still "grope" for items, terrain and furniture and it doesn't take any time. It makes it impossible to see whether you're moving, which can be a big problem with no point of reference. Walls don't show up on the map and also cause no message when they stop your movement.

Rivet-the-Zombie commented:
Yeah we really do need to retain the minimum single square of visibility in complete darkness. Trying to operate without it is pure frustration.

kevingranade commented:
Ok good points, I've seen it work, but we don't have any features to make it happen, so I'll at the very least put it off for some other PR (which is probably never TBH, not super high on my priorities).

stk2008 commented:
I agree that zero square visablity would be very frustrating IRL at least we have our senses touch mainly but that feedback in the game is not implemented.

I may as well add this here I had an idea

How about different weather effects effect vision distant eg 

Fog

So even in pure day time visabilty is reduced?

kevingranade commented:
That's very simple with this change in place, we can set the default
transparency value lower in certain weather conditions, the hard part is
actually adding the new weather type.

stk2008 commented:
:) cool well im looking forward to this being added.

Sounds really cool :)

kevingranade commented:
The issue where you get a visible line between the player and a source of illumination is a property of how light attenuation is being modeled, the intensity of light falls off at the same rate from the source to the illuminated square as it does from the illuminated square to the player, so if the player can see the illumination source, any point between the player and the source are visible.
The fix for this is implementing the inverse square law for light, so light would fall off faster than simple attenuation, yay more math :P

kevingranade commented:
I'm not totally sure what the specific error was, but the inconsistent handling of x/y values was happening in apply_light_source, where it was doing manual raytracing operations using bresenham line drawing.  Since it was unidirectional I just modified it to use the shadowcasting implementation in castLight, which is much more consistent about tracking distances etc, and faster since it only visits each square once.
There are possibly some problems remaining with cone-based light sources like headlights, which will be quite a bit more of a pain to adjust to use castLight.

kevingranade commented:
In theory all that is left is some profiling of fire handling, and tuning some light levels, getting close now.

i2amroy commented:
Definitely looking forward to this. I remember taking a stab at some light fixes a few months back and having my attempts totally ruined by the strange code. Will be nice to have a single, unified system.

kevingranade commented:
Rebalanced all the light levels, the results aren't exactly the same as they were previously in most cases, but it's pretty close and they weren't exactly fine-tuned before either.
I'm on the reading issue, and I'll be looking at the performance issue today, so this is definitely ready to start looking at for balance.

kevingranade commented:
Overhauled the code for modifying times for reading and other fine detail work, it works as intended with the new system, and now scales smoothly from 1.0x to 4.0x with light dimming.
A huge chunk of the slowdown from fires was because I switched them over to using castLight, but I forgot to update castLight for situations where they should cover a small area, now castLight accepts a threshold value that tells it to exit early.

Short of something I've missed, this is ready to go, I'm not tackling making lighting directional right now.

BevapDin commented on src/player.cpp:
```diff
@@ -13448,11 +13432,11 @@ Creature::Attitude player::attitude_to( const Creature &other ) const
     static const std::string str_bio_night("bio_night");
     const int wanted_range = rl_dist( pos3(), t );
-    bool can_see = Creature::sees( t, bresen1, bresen2 );
+    bool can_see = g->m.pl_sees( t, wanted_range );
```
`player::sees` is inherited by the npc class (and not overridden), wouldn't that make NPCs use `pl_sees` and thereby the player-visibility cache? And `pl_sees` has an explicit reference to `g->u`.

BevapDin commented:
Something is odd with the shadow casting. I can't really explain it, so here's a picture. This is in the middle of nowhere, there is only dirt and grass (and occasional shrubs) around and a *single* tree. It looks as if the tree blocks the whole 45° area that a single `castList` call would handle.

![screenschot-2015-05-25-07-49-08](https://cloud.githubusercontent.com/assets/5095435/7792449/b12557ea-02b2-11e5-945f-6a5dc0f67fa2.png)

kevingranade commented:
It slipped my mind that our inheritance between player and npc is still upside down.
I'm not exactly sure what's causing the FoV problem, but I'm pretty sure when I broke it.

kevingranade commented:
Dispatched the sight check to Creature::sees if the player in question is not THE player.
Fixed a parameter mixup that was causing the FoV thing.

BevapDin commented:
It works fine so far. One thing through: the atomic lamp and the flaming chunk of steel illuminate a larger area (about 3-8 tile radius) when on the ground, but basically nothing when in the inventory.

That seems inconsistent. It is also shown in the crafting screen: very dark vs brightly

kevingranade commented:
Preexisting issue, items on ground have item::getlight() called on them, and items in inventory have item::getlight_emit_active called on them, I bypassed and removed getlight_emit_active() since nothing ever cares if the item is active or not.

BevapDin commented:
> Preexisting issue, 

Right. It can be left for later. At this point, I would merge it, but it changes the lightning inside of buildings considerably and that is a balance issue, which I can't judge.

Here is a picture in the shelter after the normal game start. Notice the invisible inside squares (they are really invisible, no information about monsters or items there). This looks the same regardless of the weather:


![screenschot-2015-05-25-19-07-58](https://cloud.githubusercontent.com/assets/5095435/7800183/eb377152-0311-11e5-8eeb-5dfb14d0a2fd.png)

Not sure if intended, but weather does not affect the sight range during normal daytime, it only affects how many of the visible outer rows/columns are shadowy. Bad weather means you can see 40 (or so) tiles with color and 6 further tiles without color, good weather means you can see 45 tiles with color and 1 tile without color (the same effective radius).

Weather does affect the sight range during dawn: upon bad weather, dawn happens later.

On clear weather (spring, second day), during dawn, the sight range increases every 5 (or so) turns by 5 squares. It should probably be smoother (increase every 2 turns by 1 square or so).

Another thing (also preexisting, I think): after rain drops (or snow) have hidden the terrain, it is redrawn with color, even if it's in the shadowy part.

----

Are those things intended or acceptable (at least for now)?

kevingranade commented:
The interior lighting is working as intended, light is cast through the window into the building, but otherwise it's very low light and that limits your vision range.
I'll take another pass at adjusting the weather brightness, it should be restricting max sight range somewhat.
I think the dawn/dusk thing will be an easy fix as you outline, will look into it.
Precipitation, always hacky, always broken, I'll see if there's a simple fix, but if there isn't I'm not going to spend a lot of time on it.

Thanks for the testing!

kevingranade commented:
The weather thing is tougher than I thought, in order to make lowering light levels bring down visibility to where it was before, it makes the outer half of the area shadowed.

kevingranade commented:
Fixing up sunrise/sunset to be based on seconds instead of minutes, it's a continuous function that's recalculated every turn, so no idea why it was set to minutes in the first place.

kevingranade commented:
Sorted out sunrise and sunset, I think I have an avenue of attack for re-enabling weather-based FoV restriction, which is to have one component that dims the light (overcast) and another component that lowers per-square visibility, effectively like having fog or smoke on every tile.
While digging into that I realized that I never sorted out monster and NPC vision range, it'll be fairly simple to patch in a loose estimate based on light level, and I don't think it'll be necessary to do anything complicated there.

DavidKeaton commented:
is the NPC and NPC vision based on any vision cache? curious (and I'd figure so)

stk2008 commented:
Also does this mean we could stick to shadows and sneak about and avoid detection from NPC`s? :)

Rivet-the-Zombie commented:
>Also does this mean we could stick to shadows and sneak about and avoid detection from NPC`s? :)

Cheeki breeki iv damke!

Anything that gives us an advantage over those foulmouthed pink muggers is good.

illi-kun commented:
LOL, Rivet, your spontaneous Russian made me laugh :D

BevapDin commented:
Looks good. But I found another bug: start in shelter (curtains & doors closed, daytime), move into the southwest corner, spawn a zombie at the console (so you can see it). As your character is in the dark area, the zombie should not see it, however, it does. The list monsters around shows a yellow `!` in front of the monster name (which means it can see the from the Lua bindings: ), and it does indeed path straight to your character (the path is shown in the map editor).

I suspect in `Creature::sees`, you need to call `range_cur = sight_range( g->ambient_light_at( t ) )`, that is `ambient_light_at` instead of `g->light_level` because the code would use the global (outside) daylight, but the zombie is inside.

Coolthulhu commented:
Are fires supposed to be grayed out when in darkness?

Coolthulhu commented on src/lightmap.cpp:
```diff
@@ -184,7 +180,7 @@ void map::generate_lightmap( const int zlev )
                     // Only apply this whole thing if the player is inside,
                     // buildings will be shadowed when outside looking in.
-                    if (natural_light > LIGHT_SOURCE_BRIGHT && u_is_inside && !is_outside(x, y)) {
+                    if (natural_light > LIGHT_SOURCE_BRIGHT && !is_outside(x, y)) {
```
`is_outside` now has to re-get the cache every time. It would be faster to use `map_cache.outside_cache` directly (or through a trivial getter).

Coolthulhu commented:
Performance near large fires is really, really bad. And it's obviously not the fire-specific calculations, but light or visibility. Same thing happens if I spawn a large block of turrets. A 9x9 block of turrets drops my "framerate" to the point where it takes 15 seconds just to wait 5 minutes.

This happens regardless whether I can see the turrets/fire or not. Just looking around takes ~0.5 second per tile changed.

Coolthulhu commented:
Flashlight range is huge - 22 tiles (I counted). Tested in hotel's basement.
Its range wouldn't drop until it reached 16 charges, then it instantly dropped to 0.

Coolthulhu commented:
To sum up:

Major problems:
* Performance scales really badly with a large number of light sources. A single burning house is enough to make the game unplayable
* Flashlight (and presumably other light sources) has huge range until it reaches 16 charge, then has no range at all
* Sunrise/sunset is very fast at low light levels. Visibility range can easily jump 2 tiles or so in a single turn.
* Holding a small light source can kill vision radius. For example: activate full nightvision, get underground, drop all lights (get huge vision range), spawn atomic nightlight and see nothing. Even dropping it won't help until you move out of the tile.

Minor problems:
* Small light sources are grayed out when in the distance
* Light shape is sometimes a bit weird - jagged on the edges. This typically happens near fires
* "Shady" light status doesn't exist. Despite grayed out tiles, crafting menu shows only "Brightly" and "Very dark". I wasn't able to get a shady light neither by manipulating time (and thus daylight) nor by walking around light sources. Combining them both didn't help either.
* Only medium density smoke and poison gas show off the new system properly - light smoke/gas is as transparent as air (checked with debug menu 'g') and heavy smoke/gas is fully opaque
* (Not sure if bug or feature) Light-producing creatures are not seen through thick gas, but their ground tiles are
* Full night vision is really good underground. Wasn't underground supposed to be pitch black now? It probably should not work without at least a lightstrip or something like that to produce a tiny glow.

The vision through (mid-density) smoke looks nice. This could be extended to more gas types, to fires and structures, such as reinforced glass (it is air-grade transparent now - could be a bug).

kevingranade commented:
Monsters seeing too far is the issue I mentioned I'm working on right now,
and you're right that map::light_level() is the issue.
I'll take another pass over the profile data and figure out what I can do
about it, an approximate exp() function might be a big help, turns out the
proper one is super slow.
I'll re-enable the 'high visibility for light sources' logic, it was so
subtle I didn't realize it was there at first, light sources are applied to
map::cache::sm (for source map I guess) as well as map::cache::lm, and
there are a handful of overrides that use it scattered around.
I tweaked all the light source numbers to match previous behavior, not sure
what the disconnect is.
I'm not sure that visibility changing by two tiles in a turn at dawn or the
end of dusk is a major problem.  I'll try wrapping it in the logistic curve
function, I think that'll do what we want.
Holding a small light source killing night vision is intended, it has a
light level threshold where it deactivates instead of simply being a boost
to sight range.  It could be more gradual, but that can wait.
When the squares of light-producing monsters is visible through smoke, is
it labeled with the terrain type, or 'bright'?  If the latter it's
intended, if the former it's a bug.
I'd prefer to leave underground nightvision where it is, didn't it work
underground before?

NPC and monster vision isn't cached as heavily as pc vision because it's
only used on-demand to spot-check visibility of individual targets instead
of being applied to every square on the map.  It does use the player caches
when they can.

Coolthulhu commented:
> Holding a small light source killing night vision is intended, it has a light level threshold where it deactivates instead of simply being a boost to sight range.

Shouldn't such a tiny light source be hidden in a pocket or something? Lightstrips and similar items are pretty big anti-tedium mechanics. Having to drop them not to kill own vision radius sounds rather painful.

> When the squares of light-producing monsters is visible through smoke, is it labeled with the terrain type, or 'bright'?

Terrain type - it also draws the tile as if it was visible. In case of turret, it draws a 3x3 square of tiles. The creature itself isn't visible.

> I'd prefer to leave underground nightvision where it is, didn't it work underground before?

I got the impression that pitch darkness means no nightvision. If it's a decision to keep it working, then it's not a problem.

Zireael07 commented:
FNV worked underground brilliantly before, @Coolthulhu, so I see no reason to change it. Maybe in another PR if someone feels it's a balance problem.

Coolthulhu commented:
Yeah it did, it's just that I got the impression that NV was supposed to stop working with pitch black darkness. This would make sense from realism point of view, though wouldn't be too good for gameplay.

kevingranade commented:
Re light killing nightvision, I'd buy changing the threshold to exclude
some of the dimmer lights, but 'stash it in a pocket' isn't something I
want to mess with, sounds cumbersome.
You may notice I originally had full dark be really really full dark, but
was convinced to change it as we don't have a 'feel to explore and remember
terrain' type mechanic, which is one of the things that makes that a
reasonable thing to do.

DavidKeaton commented:
NVD can also have an IR attachment providing light in 0 light areas.

kevingranade commented:
Probably fixed creature sight range, it passes BevapDin's use case anyway, but I haven't tested it well yet.
Thanks for the pointer about truncation, made sunlight() return a float.
Bumped up the threshold that prevents nightvision, now nothing dimmer than an oil lamp spoils night vision.

Coolthulhu commented:
> 'stash it in a pocket' isn't something I want to mess with, sounds cumbersome.

I mean, in current master some very weak lights (lightstrip and atomic nightlight) can be safely carried by the player and used to craft and read. They don't produce enough light to reveal the player nor to see the next tile - essentially they're just stored in the pocket and taken out to craft/read.

I understand that the NV kill is mostly there to prevent it boosting flashlights, which makes sense and provides some balancing to NV.
But "realistically" very weak light sources such as glowsticks would probably boost NV, especially the weak pseudo-NV that a regular human could have.

Not a high priority thing, but that NV killing threshold could be changed to depend on current mutation level. Or even continuously depend on both light level and mutation level.

Coolthulhu commented:
> "jagged" light borders?

Not sure how to explain properly.

When a large, irregular source of light (such as a fire) is moderately far from the player, the lit up area is only very roughly round. It has "triangles" extending roughly perpendicularly to the player<->fire line.
Not a high priority thing, but it may suggest some sort of a bug or strong order dependency.

I may not be fixable nor actually important.

kevingranade commented:
> I mean, in current master some very weak lights (lightstrip and atomic nightlight) can be safely carried by the player and used to craft and read. They don't produce enough light to reveal the player nor to see the next tile - essentially they're just stored in the pocket and taken out to craft/read.

Whatever your rationale, bumping the threshold should do it for the time being, you can have any of these on you and not be affected:
cigarette, cigar, joint, lightstrip, atomic nightlight, glowstick, candle, refillable lighter, charcoal kiln
Basically what you were saying, anything that actually contributes to vision range.

kevingranade commented:
I'm a little stuck on handling light source visibility, the old code basically says if there's a light source in your line of sight, you can see it, period.  I'm trying to figure out how to translate that into letting it get obscured while still boosting it a bit.

kevingranade commented:
Meh, I think I'll make it relatively unlimited for now, i.e.
```
return map_cache.seen_cache[tx][ty] > LIGHT_TRANSPARENCY_SOLID + 0.1 &&
    ( map_cache.seen_cache[tx][ty] * map_cache.lm[tx][ty] >
      g->u.get_vision_threshold( map_cache.lm[g->u.posx()][g->u.posy()] ) ||
      map_cache.sm[tx][ty] > 0.0 );

```

kevingranade commented:
Here's what the profile breaks down as for running around in circles in a mall:
```
+   79.73%     0.00%  cataclysm  [unknown]            [.] 0xf656258d4c544155
+   79.73%     0.00%  cataclysm  libc-2.20.so         [.] __libc_start_main
+   79.73%     0.00%  cataclysm  cataclysm            [.] main
+   62.81%     0.01%  cataclysm  cataclysm            [.] game::do_turn
+   40.48%     0.22%  cataclysm  cataclysm            [.] map::build_map_cache
+   36.55%     3.62%  cataclysm  cataclysm            [.] map::generate_lightmap
+   28.40%     0.45%  cataclysm  cataclysm            [.] map::apply_light_arc
+   26.24%    19.23%  cataclysm  cataclysm            [.] map::apply_light_ray
+   25.12%     0.00%  cataclysm  cataclysm            [.] game::draw
+   24.05%     0.00%  cataclysm  cataclysm            [.] game::draw_ter
+   24.04%     0.00%  cataclysm  cataclysm            [.] game::draw_ter
+   15.89%     0.00%  cataclysm  cataclysm            [.] game::opening_screen
+   10.62%     0.00%  cataclysm  cataclysm            [.] game::setup
+   10.50%     0.00%  cataclysm  cataclysm            [.] game::load_world_modfiles
+    5.84%     0.28%  cataclysm  cataclysm            [.] map::draw
+    5.58%     5.51%  cataclysm  libm-2.20.so         [.] __ieee754_exp_sse2
+    5.40%     0.00%  cataclysm  cataclysm            [.] game::load_data_from_dir
+    5.40%     0.00%  cataclysm  cataclysm            [.] game::load_core_data
+    5.40%     0.00%  cataclysm  cataclysm            [.] DynamicDataLoader::load_data_from_path
+    5.35%     0.00%  cataclysm  cataclysm            [.] JsonIn::get_object
+    5.30%     0.11%  cataclysm  cataclysm            [.] JsonObject::JsonObject
+    5.10%     0.00%  cataclysm  cataclysm            [.] DynamicDataLoader::finalize_loaded_data
+    5.08%     0.00%  cataclysm  cataclysm            [.] game::load
+    5.03%     0.00%  cataclysm  cataclysm            [.] map::loadn
``
What jumps out of course is:
+   26.24%    19.23%  cataclysm  cataclysm            [.] map::apply_light_ray
which happens in two places, vehicle headlights and casting light through windows.
In this case I'm pretty sure it's a bunch of windows, but it's just as bad if there are a large number of vehicles around.
One good solution for this is making apply_light_arc() use castLight to do the heavy lifting, it's far, far cheaper to run than the raycasting approach apply_light_arc() uses, unfortunately I'm having a lot of trouble taking an input angle and origin and chopping it up into the appropriate octants and sub-octants that castLight() needs to operate on.
If I make light projecting through windows spread in a 90 degree angle instead of the 60 it uses now, it'd be much easier, but that still leaves cars being problematic since they can be rotated arbitrarily.

kevingranade commented:
After a kind of ridiculous amount of overhauling, it looks like this now:
```
+   80.40%     0.00%  cataclysm  [unknown]                     [.] 0xeca6258d4c544155
+   80.40%     0.00%  cataclysm  libc-2.20.so                  [.] __libc_start_main
+   80.40%     0.00%  cataclysm  cataclysm                     [.] main
+   66.58%     0.04%  cataclysm  cataclysm                     [.] game::do_turn
+   27.54%     0.00%  cataclysm  cataclysm                     [.] game::draw
+   25.87%     0.00%  cataclysm  cataclysm                     [.] game::draw_ter
+   25.86%     0.00%  cataclysm  cataclysm                     [.] game::draw_ter
+   17.66%     0.25%  cataclysm  cataclysm                     [.] map::build_map_cache
+   14.54%     3.14%  cataclysm  cataclysm                     [.] map::generate_lightmap
+   12.88%     0.00%  cataclysm  cataclysm                     [.] game::opening_screen
+   10.27%     0.41%  cataclysm  cataclysm                     [.] map::draw
+    9.77%     4.60%  cataclysm  cataclysm                     [.] game::update_scent
+    9.57%     0.02%  cataclysm  cataclysm                     [.] map::apply_directional_light
+    8.76%     0.00%  cataclysm  cataclysm                     [.] game::setup
+    8.67%     0.00%  cataclysm  cataclysm                     [.] game::load_world_modfiles
+    8.11%     0.00%  cataclysm  cataclysm                     [.] map::process_active_items
+    8.09%     0.04%  cataclysm  cataclysm                     [.] map::process_items<bool (*)(item_stack&, std::_List_iterator<item>&, tripoint const&, std::string)>
+    7.37%     0.83%  cataclysm  cataclysm                     [.] map::update_visibility_cache
+    7.06%     0.04%  cataclysm  cataclysm                     [.] map::process_items_in_submap<bool (*)(item_stack&, std::_List_iterator<item>&, tripoint const&, std::string)>
+    6.26%     0.02%  cataclysm  cataclysm                     [.] process_map_items
+    6.23%     0.03%  cataclysm  cataclysm                     [.] process_item<std::_List_iterator<item>
+    5.86%     1.28%  cataclysm  cataclysm                     [.] map::apparent_light_at
+    5.11%     0.01%  cataclysm  cataclysm                     [.] map::scent_blockers
+    5.05%     0.41%  cataclysm  cataclysm                     [.] player::update_bodytemp
```
Turns out the two biggest contributors were regenerating the map cache twice a turn so monsters get up to date vision data (the fix for the "turret shoots me though closed door" issue a while back) and the early termination code for applying dynamic lighting, it wasn't applying the inverse square decay I added to light projection, so it was continuing to calculate light project far after it wasn't having an effect any more.

I have another large optimization to do that should fix the scenario where there are a lot of fires uselessly casting light through each other, and I strongly suspect that I fixed the "jagged light borders" issue Coolthulu noticed.  There is some code that tries to suppress light emission when light sources are adjacent to each other, and at some point I messed up the directions it was deciding to disable.

kevingranade commented:
BTW, the optimization I referenced is taking the existing optimization explained around lightmap.cpp:753 and generalize it by keeping a map of all the light sources, and terminating light casting if the current light ray is not as strong as the light source it encounters.  Additionally I'll treat naturally illuminated squares as light sources for this purpose, meaning that at least in the daytime dynamic lighting should frequently terminate earlier.

kevingranade commented:
I forgot, there was another sizeable slowdown, which was the std::function I was passing to castLight, turns out dispatching to one of those has non-trivial overhead and it was noticeable in a very tight loop like castLight.  That's why I went for the more complicated solution of passing methods as explicit template parameters.

kevingranade commented:
I took a look at the "tons of fire" use case, and I'm not sure it needs any help from the lightmap end.  I lit up a mall, hundreds of tiles of fire, and it sure slowed down a ton, but this is what I'm seeing in profiling:
```
+   85.08%     0.00%  cataclysm  cataclysm            [.] main
+   85.08%     0.00%  cataclysm  libc-2.20.so         [.] __libc_start_main
+   85.08%     0.00%  cataclysm  [unknown]            [.] 0x06b6258d4c544155
+   80.88%     0.01%  cataclysm  cataclysm            [.] game::do_turn
+   52.42%     0.01%  cataclysm  cataclysm            [.] map::process_fields
+   50.13%     2.89%  cataclysm  cataclysm            [.] map::process_fields_in_submap
+   18.61%     0.04%  cataclysm  cataclysm            [.] map::build_map_cache
+   18.04%     2.07%  cataclysm  cataclysm            [.] item::made_of
+   12.42%     0.00%  cataclysm  cataclysm            [.] game::draw
+   12.30%     0.00%  cataclysm  cataclysm            [.] game::draw_ter
+   12.29%     0.00%  cataclysm  cataclysm            [.] game::draw_ter
+   10.68%     0.85%  cataclysm  cataclysm            [.] item::made_of
+    9.67%     0.26%  cataclysm  cataclysm            [.] map::generate_lightmap
+    6.39%     1.09%  cataclysm  cataclysm            [.] item::volume 
+    5.90%     1.01%  cataclysm  cataclysm            [.] std::vector<std::string, std::allocator<std::string> >::operator=
+    5.43%     0.03%  cataclysm  cataclysm            [.] map::apply_light_source
+    5.14%     0.00%  cataclysm  cataclysm            [.] map::build_seen_cache
+    5.04%     0.71%  cataclysm  cataclysm            [.] map::process_fields_in_submap(submap*, int, int, int)::{lambda(field_entry*, tripoint const&, field_id, int, int)#2}::operator()  
```
In other words, it's field processing that's dominating things now, I'm guessing switching field lighting over to castLight, plus the other general optimizations I did in there resolved any issues you saw before.  Lightmap operations are certainly contributing, but the next step for optimizations is really damn invasive and messy, and I'd rather not make it if I don't absolutely have to.
It looks like item::made_of and item::volume are dominating, at a combined 25% of CPU load....
Now that I'm looking, item::made_of(std::string) is fairly terrible, and it doesn't have to be, will PR something to address that shortly.

DavidKeaton commented:
Now I need to look at `made_of` to see why. ;-)

kevingranade commented:
I have a 10% speedup for the "huge pile of fires" scenario, and I'm doing a test build of a rebase, after that I'll push and this should maybe be good to go.

kevingranade commented:
Nothing in particular is broken, but I just ran out of time to rebuild and test after rebasing, so I don't feel right calling it "ready", if you feel like testing anyway and it checks out though, it should be mergeable.

BevapDin commented on src/mtype.cpp:
```diff
@@ -64,6 +64,11 @@ void mtype::set_flag(std::string flag, bool state)
 
+bool mtype::has_material( const std::string &material ) const
+{
+    return std::find( mat.begin(), mat.end(), material ) == mat.end();
```
Should be `!=`, shouldn't it? It's that way in `item::made_of`.

kevingranade replied:
Hah, forgot to check that fix in.  Thanks for catching it.

BevapDin commented on src/item.cpp:
```diff
@@ -2730,21 +2730,17 @@ bool item::is_two_handed(player *u)
-    } else if (is_corpse()) {
+        return null_material;
+    } else if( is_corpse() ) {
         // Corpses are only made of one type of material.
```
They aren't anymore.

BevapDin commented:
There is some strange effect with cars during snowstorms (does not happen in clear weather).

Out in the open field, in the middle of a submap, spawn a car. Open and close all doors, walk around the car (this should all happen one the very same submap, so no map-shifting). Apparently some parts of the car count as opaque.

This effect goes away as soon as the map shifts. Opening *or* closing a car door recreates the effect.

![screenschot-2015-06-03-17-06-47](https://cloud.githubusercontent.com/assets/5095435/7963521/600f2f3a-0a13-11e5-906f-63f4083c9f72.png)

DavidKeaton commented:
Should be a relatively easier fix then, yeah?

Coolthulhu commented:
Minor:
* Weather effects don't show up in nighvision-visible areas
* Near areas with flickering light such as fires, nightvision constantly flickers on and off
* Car headlights light up a small area - both short and narrow
* Doesn't compile because of a missing comma in mtype.cpp

Moderate:
* The snowstorm bug BevapDin discovered and described. It reproduces very reliably
* Flashlights provide full light until 16 charges, when they instantly stop working altogether

Bugs unrelated to this PR:
* You can close car door on the tile you occupy
* Houses no longer collapse in fire (my fault)

DavidKeaton commented:
The vehicle door bug has been around a while. I won't lie, I use it quite often. ;-)

kevingranade commented:
nnnng... thought I had all the shadowcasting bugs sorted out, looks like this is one that only happens when it transitions from one transparency level to another.
The trigger is that squares in the vehicle that count as inside aren't affected by the adjustment to ambient transparency from weather (in other words, precipitation isn't blocking your view, that part is intended).  What happens is it recursively shadowcasts the span before, in, and after the square with different transparency seperately, but the edges aren't matching up, so it develops a gap that is not cast.  The same thing happens if you just throw a smoke grenade, but I never noticed because I expected part of my view to be blocked then, but if you examine the light levels around where it dims out (with debug->g), it doesn't taper off to black, it just suddenly goes from full to dark.
Not sure how quick a fix is going to be, that's some damn tricky code :/

kevingranade commented:
Side note, I'm guessing that the effect flipping on and off is due to vehicle caching effects or something, since the affected squares count or inside or not and therefore trigger the casting bug or not depending on whether the doors or open or not.

kevingranade commented:
Ow my head, but I think I have it sorted out.
BTW, this is the explanation for how recursive shadowcasting works: http://www.roguebasin.com/index.php?title=FOV_using_recursive_shadowcasting
Except I've heavily modified it since then :/

I definitely still need to tweak headlight and possibly flashlight values, and at least take a look at making night vision drop off gradually instead of suddenly disappear.

kevingranade commented:
Just a little detail of what that last change did:
As the cells are being processed, the shadowcaster is calculating lines to the corners of cells like this diagram from the RogueBasin article:
```
 +---+xxxxx#####  x = first blocking cell
 |   |xxxxx#####  a = point that 'brushes' by to
 |   |xxxxx#####      the left of the blocking cell
 |   |xxxxx#####
 +---axxxxx#####
 +---++---++---+
 |   ||   ||   |
 |   ||   ||   |
 |   ||   ||   |
 +---++---++---+

 Fig.4d Zoom in on the first blocking cell
```
When transitioning from one cell to the next (that have differing transparency), it needs to recursively shadowcast using the "previous start angle" (see below) to point 'a' in this diagram.
The tricky part is the definition of "previous start angle".  In the previous algorithm, it was always the corner opposite 'a' as outlined in this diagram:
```
 ##########aoooo  o = first non-blocking cell
 ##########o   o  a = point that 'brushes' by to the
 ##########o   o      right of the blocking cell
 ##########o   o
 ##########ooooo
 +---++---++---+
 |   ||   ||   |
 |   ||   ||   |
 |   ||   ||   |
 +---++---++---+

 Fig.4g Zoom in on the first non-blocking cell
```
With our code though, we additionally have to handle the case where it's transitioning from one transparent cell to another transparent cell with different opacity, in that case the start angle needs to be a third corner of the cell, this one:
```
 ######xxxxooooo  o = transparent cell with different opacity
 #######xxxo   o  a = point that 'brushes' by to the
 ########xxo   o      nearby corner of the previous cell
 #########xo   o  x = portion of previous cell we
 ##########aoooo      shadowcast through.
 +---++---++---+
 |   ||   ||   |
 |   ||   ||   |
 |   ||   ||   |
 +---++---++---+

 Fig.4x Zoom in on a second transparent cell.
```
As you can see, the end angle of the previous span is the same as the start angle of the next span, which must be the case to avoid gaps.

If you want an even bigger headache, take a look at my old code that mixes portals into this.

DavidKeaton commented:
Good article, read it a while ago, p good.

kevingranade commented:
Nevermind, that screwed something up, I can see through walls now >_<

kevingranade commented:
Un-screwed maybe, this code is such a pain in the ass.

BevapDin commented on src/item.cpp:
```diff
@@ -4033,14 +4020,12 @@ int item::getlight_emit() const {
-            lumint = ( type->light_emission * chargedrop * charges ) / maxcharge;
+        // Falloff starts at 1/5 total charge and scales linearly from there to 0.
+        if( maxcharge > 0 && charges > 0 && maxcharge / charges > 5 ) {
+            lumint *= charges * 5 / maxcharge;
```
I thinks this is responsible for the flashlight bug Coolthulhu mentioned. When charges is 17 (maxcharges is 100), the if-condition is not met, but as soon as it drops to 16 charges, it is. And then `charge * 5 / maxcharges` evaluates to 0 (a `long`) all the time.

kevingranade replied:
oh hell, that looks like it, testing now.

kevingranade commented:
Knocked out several of the remaining issues, rebased, out for the night.
I should be able to handle the animation based issues relatively quickly, it should just be a matter of adding an appropriate check and refusing to draw on tiles the player can't see.

DavidKeaton commented:
What issues are there left, @kevingranade ?

kevingranade commented:
Two major things, make animations (weather and SCTs) respect vision, and
make nightvision cancellation gradual so it doesn't toggle on and off
jarringly.

kevingranade commented:
Holy crap this has been so much more work than expected >_<
All known issues are addressed though, please take a look.

kevingranade commented:
As a bonus I noticed in final testing that monsters were showing up outside the area that should be visible, and fixed that.
The night vision toggling is somewhat incremental now, but there are generally only a few squares between total darkness and bright light, so I can't think of a way to make it very gradual without messing with a ton of other things.

Coolthulhu commented:
Some weird jaggies, might be just smoke:
http://postimg.org/image/pkjl0sgij/

Coolthulhu commented:
Weirder jaggies - also smoke?
http://postimg.org/image/kuf81f38b/

Any better image upload site?

Coolthulhu commented:
OK, this one is pretty obviously a bug:
http://postimg.org/image/x2vnk3u8l/

The fire visible is a 2x2 fire on non-transparent surface (walls). Except for that 2x2 fire+wall block, there is no other non-transparent tile around.

Coolthulhu commented:
After one of the fires decayed:
![after](https://imagizer.imageshack.us/v2/348x486q90/537/Y8is7u.png)

Coolthulhu commented:
Before:
![before](https://imagizer.imageshack.us/v2/272x381q90/905/T5Q28k.png)
After (only change is the decreased fire density on that single tile):
![after](https://imagizer.imageshack.us/v2/276x368q90/901/UUUUMB.png)

Coolthulhu commented:
Major:
* Near-sighted mutation causes total blindness in some ways (everything is seen as "darkness"), but still allows seeing creatures and items. Seems that `seen` is not changed, while visibility cache is.

Minor:
* The weird effect from fires on walls above
* Boomer effect causes everything to be described as pink blur, but still draws adjacent zeds and describes their actions. Presumably the same bug as with near-sighted

Not a result of this PR:
* Chicken walker that takes a lot of damage leaves what is presumably poultry

kevingranade commented:
Found the nearsighted issue, since player::sight_impaired() returned a boolean, I assumed it meant they were completely blind, otherwise what's the point right?  Nope it means it's reduced by some unspecified amount.
I'm a little suspicious that the fire effect is from an optimization that avoids casting light through an adjacent square if that square has a light source as bright as the current one, I'm trying just turning it iff to see if the effect clears up.
Boomer effect is roughly supposed to do that, can only see adjacent stuff?  I'd personally be fine with 0 visibility when boomered.

Coolthulhu commented:
> I'm a little suspicious that the fire effect is from an optimization that avoids casting light through an adjacent square if that square has a light source as bright as the current one, I'm trying just turning it iff to see if the effect clears up.

It's rather minor and rare. If the optimization is important, it should probably get the priority.

> Boomer effect is roughly supposed to do that, can only see adjacent stuff?

It is supposed to let you see everything adjacent to you. But here it only shows adjacent monsters, but doesn't describe them when 'l'ooking at them. You can still see with with 'V' monster listing and get their messages ("boomer spews bile!"), just not 'l'ook at them.
Zero visibility would make them really annoying to deal with, because bile duration isn't low and we have no way of telling the player whether a move succeeded, so it would result in keypress counting. Would make eye protection as vital as mouth protection for early survivors, but in a way more annoying way.

kevingranade commented:
Ok, I see what you're saying with the boomer issue, I think I have a fix for the nearsightedness thing, and I don't have a clue what's going on with the fire.  Looking at the old code, it looks like it does the same thing, but it looks different there.

Coolthulhu commented:
I'd say it would be OK to merge if the fire issue was the only one left.
It seems only to happen when the light sources are on very low transparency tiles. I'll test it with smoke to make sure, but even if it affected both walls and smoke, it's still rather small.

I'd suspect the whole "don't lit up walls from the other side" may be causing it, but I didn't actually review the code enough to say if it is the case.

kevingranade commented:
I was looking into that, but never got a workable solution, and didn't want to delay this one with that fix.
I'm also inclined to just leave the fire thing for now, opening an issue to track it.  I'll update when I have the other two issues sorted out, which I think will be soon.

kevingranade commented:
And there we are, you were right that it was the same issue.

Coolthulhu commented:
Can't see fire sources outside field of view while near-sighter or boomered. This could be really problematic in labs, especially now that walls won't be lit up from the other side.

Coolthulhu commented:
Very minor, I guess:
Labs have those containment cells. They were lit up before, now they aren't.

kevingranade commented:
Can you be more specific about the containment cells?  There are a bunch of different kinds of monster containment in labs, and I'm not seeing any lighting in them.

Coolthulhu commented:
Oops, the containment cell thing must have been my misinterpretation. I remembered the cells being lit up, but apparently they weren't even in master.

kevingranade commented:
There are some consoles scattered around, and some of the monsters glow, either of those could give that impression.

Coolthulhu commented:
I'm compiling, meanwhile I'll prepare a list of test to conduct, so if (when) I miss something, it will be more obvious where to search.

* [ ] Near-sighted, boomered and underwater - all of: look, 'V' and creature action descriptions ("boomer spews bile!"
* [ ] Fire burning down a building (just the light obviously, I still haven't fixed that no floor burning thing)
* [ ] Vehicle: inside, outside, driving. Walking around, opening/closing non-opaque doors, opaque doors and curtains
* [ ] Light-producing items. Carried and on the ground. Dimming flashlights, nightvision and low-light items
* [ ] Monsters producing light - turrets, burning mons
* [ ] Monster vision - total darkness, smoke, flashlight. Seeing the player and seeing other monsters
* [ ] Lights in the darkness, fires in snowstorm
* [ ] Snowstorm and day/night
* [ ] Vehicle lights
* [ ] Vehicle lights and monster vision - both player vision and infighting
* [ ] Daylight increasing and decreasing
* [ ] Moonlight
* [ ] Entering insides of a building in day, night and at dawn

Coolthulhu commented:
Could be major, could be intended:
* In summer 14 (turn 403199), moonlight produces 19 light per tile, which gives 44 tiles of vision without any nightvision. Moonlit summer nights can be quite bright, but I'm not sure if they can be THIS bright
* Floodlight produces 50 tiles of light when the player stands directly under it. Essentially daylight

Minor (I wouldn't hold up a merge for any of those), some could be intended:
* Burning walls produce very directional light
* Nightvision change at the edge of light is rather abrupt for high and full NV
* Tiles seen as "brightly" for crafting can still be drawn without colors. May be just rafting/display discrepancy
* In summer 14 (turn 403199), moonlight produces 19 light per tile, in the next turn after it (autumn 1) it produces 3.7. The change is huge and very abrupt
* "Bright light"s can be visible in opaque vehicles, but only from afar or when vision is impaired. For example, luxury RV's aisle lights
* Glowsticks are useless except for reading/crafting now
* Turrets produce tons of light
* Snowstorm has huge effect on weak light, especially moonlight

Observations (not bugs or not caused by the PR):
* Being underwater should wash down boomer bile
* Non-aggressive animals are attracted to sound, even those that are afraid of it
* Smoke is very "binary" - only medium-density smoke makes full use of the new light system
* Fire density (or at least its light production) could use more granularity. It "sparks up" way too much, almost like it was flashing instead of burning
* I overdid the zombie aggression on small animals nerf a tiny bit - they should attack dogs, because dogs can reanimate
* Curtains are welded to windshields. Foldable seats can be installed on curtains.

Recompiling in TILES now.
I'd merge it with the current bugs, unless tiles build is broken. If no one tells me not to merge, I'll merge it in current state.
Waiting for opinions until 5 GMT, then I'm merging if no one objects.

kevingranade commented:
Moonlight doesn't sound quite right, but I don't think it's worth holding anything up for, if anything the issue there is I think you can read in that light level.
Floodlight is intended to be that bright.
Fire lighting doesn't really have any requirements, so I don't much care if it changes compared to previously.
That's the best I could do with the nightvision, I can't think of anything better to do with it.
I don't recall if I was even trying to get "lit" tiles to equal "can read" light level.  I don't mind if they don't match.
The visible bright light at a distance inside an opaque vehicle is a bug, but let's just open an issue for it, this is tiring ;_;
Glowsticks are supposed to be visible at a distance and let you sort-of read when wielding them, I think that's what they do now.
Turrets might need dialing down, I don't recall if I specifically tuned them, but let's follow up later.
Snowstorm is supposed to be super harsh on visibility.

Coolthulhu commented:
Glowsticks in master give 2 tile visibility, here they don't.
Snowstorm is only really harsh on weak light. In full daylight, it doesn't do all that much. I guess daylight is so strong it is limited primarily by the bubble.

Coolthulhu commented:
OK then, tested tiles. I said I'll wait until 5 GMT, but I that was mostly in case no one replies. I'll merge now.

Brace for inevitable bugs report though.

Coolthulhu commented:
Also, congrats there Kevin. This PR was quite huge.

kevingranade commented:
Yes, it was huge and tiring, I've probably put 50+ hours into this sucker.
Thank you so much for the thorough testing, if I'd just pushed when I thought it was ready it would have probably taken ages to sort out all the things you caught.

stk2008 commented:
This is sweeet.

Thanks I shall play this later when I get back..if I start playing now I think I will miss my appointment so better wait :P

DavidKeaton commented:
nice job @kevingranade ! glad this PR is in now! very nice work. :-)

DavidKeaton commented:
playing it now, this is great!
